File: Full Stack Develop/Programming Language/TypeScript/src/log.ts
/**
 * A minimal printing helper: every demo uses it to print "expression → result" aligned,
 * so a single run shows the real behavior of a language feature without memorizing docs.
 * The language projects (TypeScript / Go / Python) all provide the same title / note / show.
 */
const C = {
  reset: '\x1b[0m',
  dim: '\x1b[2m',
  bold: '\x1b[1m',
  cyan: '\x1b[36m',
  green: '\x1b[32m',
  yellow: '\x1b[33m',
  red: '\x1b[31m',
} as const;

/** A section heading */
export function title(text: string): void {
  console.log(`\n${C.bold}${C.cyan}━━ ${text} ${'━'.repeat(Math.max(0, 56 - text.length))}${C.reset}`);
}

/** An explanatory line (a dimmed comment starting with #) */
export function note(text: string): void {
  console.log(`  ${C.dim}# ${text}${C.reset}`);
}

/** Print "expression → result", aligned as expr  →  value */
export function show(expr: string, value: unknown): void {
  console.log(`  ${C.green}${expr.padEnd(44)}${C.reset}${C.dim}→${C.reset} ${format(value)}`);
}

export function warn(text: string): void {
  console.log(`  ${C.yellow}! ${text}${C.reset}`);
}

export function error(text: string): void {
  console.log(`  ${C.red}✗ ${text}${C.reset}`);
}

export const sleep = (ms: number): Promise<void> => new Promise((r) => setTimeout(r, ms));

function format(v: unknown): string {
  if (v === null) return `${C.dim}null${C.reset}`;
  if (v === undefined) return `${C.dim}undefined${C.reset}`;
  if (typeof v === 'string') return `"${v}"`;
  if (typeof v === 'bigint') return `${v}n`;
  if (v instanceof Map) return `Map(${JSON.stringify([...v.entries()])})`;
  if (v instanceof Set) return `Set(${JSON.stringify([...v.values()])})`;
  if (Array.isArray(v)) return `[${v.map(format).join(', ')}]`;
  if (typeof v === 'object') return JSON.stringify(v);
  return String(v);
}

File: Full Stack Develop/Programming Language/Rust/src/log.rs
//! The Rust version of the minimal printing helper: prints "expression → result" aligned.
//! Corresponds to TypeScript's src/log.ts, Go's internal/logx, and Python's demos/log.py.
//!
//! show uses `{:?}` (Debug) to format any value uniformly: strings get quotes, numbers print raw,
//! and collections print as [..], matching the display style of the other three languages exactly.

use std::fmt::Debug;

const RESET: &str = "\x1b[0m";
const DIM: &str = "\x1b[2m";
const BOLD: &str = "\x1b[1m";
const CYAN: &str = "\x1b[36m";
const GREEN: &str = "\x1b[32m";

/// Print a section heading.
pub fn title(text: &str) {
    let n = text.chars().count();
    let line = "━".repeat(if n >= 40 { 0 } else { 40 - n });
    println!("\n{BOLD}{CYAN}━━ {text} {line}{RESET}");
}

/// Print an explanatory line (a dimmed # comment).
pub fn note(text: &str) {
    println!("  {DIM}# {text}{RESET}");
}

/// Print "expression → result", aligned as expr → value.
pub fn show<T: Debug>(expr: &str, value: T) {
    let n = expr.chars().count();
    let pad = " ".repeat(if n >= 44 { 1 } else { 44 - n });
    println!("  {GREEN}{expr}{pad}{RESET}{DIM}→{RESET} {value:?}");
}
